### PR TITLE
Refactoring series focused on scala style

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

**Comment by chibenwa** [comment]
Screenshot from 2023-02-23 08-49-54

Otherwize +1, the extension pattern looks cool to me.

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — COMMENTED** [review]
Read it.

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala [hunk 1]
```diff
@@ -108,8 +109,10 @@ class GlobalRateLimit @Inject()(rateLimiterFactory: RateLimiterFactory) extends
   private var keyPrefix: Option[KeyPrefix] = _
 
   override def init(): Unit = {
-    val duration: Duration = parseDuration()
-    val precision: Option[Duration] = parsePrecision()
+    val duration: Duration = getMailetConfig.getDuration("duration")
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala:117]: scoping the DurationOps import to the init method would reduce compiler work

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala [hunk 2]
```diff
@@ -33,11 +33,15 @@ import scala.util.Try
 case class KeyPrefix(value: String)
 
 object ConfigurationOps {
+
+  implicit class OptionOps(mailetConfig: MailetConfig) {
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:37]: Why not make a MailetConfigOps with all the extended behavior ? while I am all for interface segregation this pattern results in translating
```
 implicit class DurationOps(mailetConfig: MailetConfig) {
    def getDuration(key: String): Option[Duration] = Option(mailetConfig.getInitParameter(key))	    def getDuration(key: String): Option[Duration] = new OptionOps(mailetConfig).getOptionalString(key)
      .map(string => DurationParser.parse(string, ChronoUnit.SECONDS))	      .map(string => DurationParser.parse(string, ChronoUnit.SECONDS))
  }	  }
```
then

```
  val precision: Option[Duration] = getMailetConfig.getDuration("precision")
```
gets  translated to 
```
  val precision: Option[Duration] = new DurationOps(getMailetConfig).getDuration("precision")
```
this requires 2 different implicit resolutions, 2 instanciations and seems a bit wasteful

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala [hunk 3]
```diff
@@ -36,12 +36,22 @@ object ConfigurationOps {
...
+    def getOptionalLong(key: String): Option[Long] = Option(mailetConfig.getInitParameter(key)).map(_.toLong)
   }
 
   implicit class DurationOps(mailetConfig: MailetConfig) {
     def getDuration(key: String): Option[Duration] = mailetConfig.getOptionalString(key)
       .map(string => DurationParser.parse(string, ChronoUnit.SECONDS))
   }
+
+  implicit class SizeOps(mailetConfig: MailetConfig) {
+    def getOptionalSize(key: String): Option[org.apache.james.util.Size] = mailetConfig.getOptionalString(key)
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:49]: same argument as previous comment

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala [hunk 4]
```diff
@@ -48,62 +49,48 @@ object ConfigurationOps {
...
         case "" => throw new IllegalArgumentException(s"'$key' field cannot be empty if specified")
-        case s => s
+        case s  => s
       }
       .map(org.apache.james.util.Size.parse)
   }
 }
 
+object EntityType {
+  def extractRules(entityType: EntityType, duration: Duration, mailetConfig: MailetConfig): Option[Rules] = (entityType match {
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:64]: is it expected that adding operations (new functions) will be more frequent than adding new entity types ?

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala [hunk 5]
```diff
@@ -32,10 +32,10 @@ import org.reactivestreams.Publisher
 import reactor.core.scala.publisher.{SFlux, SMono}
 
 case class GlobalKey(keyPrefix: Option[KeyPrefix], entityType: EntityType) extends RateLimitingKey {
-  val globalPrefix: String = "global"
-
-  override def asString(): String = keyPrefix.map(prefix => s"${prefix.value}_${entityType.asString}_$globalPrefix}")
-    .getOrElse(s"${entityType.asString}_$globalPrefix")
+  override val asString: String = {
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala:35]: :+1:

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala [hunk 6]
```diff
@@ -142,7 +142,7 @@ class GlobalRateLimit @Inject()(rateLimiterFactory: RateLimiterFactory) extends
 
   private def createRateLimiter(rateLimiterFactory: RateLimiterFactory, entityType: EntityType, keyPrefix: Option[KeyPrefix],
                                 duration: Duration, precision: Option[Duration]): GlobalRateLimiter =
-    GlobalRateLimiter(rateLimiter = EntityType.extractRules(entityType, duration, getMailetConfig)
+    new GlobalRateLimiter(rateLimiter = EntityType.extractRules(entityType, duration, getMailetConfig)
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala]: :+1:

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala [hunk 7]
```diff
@@ -140,12 +145,17 @@ class GlobalRateLimit @Inject()(rateLimiterFactory: RateLimiterFactory) extends
...
   }
 
+
   private def createRateLimiter(rateLimiterFactory: RateLimiterFactory, entityType: EntityType, keyPrefix: Option[KeyPrefix],
                                 duration: Duration, precision: Option[Duration]): GlobalRateLimiter =
-    new GlobalRateLimiter(rateLimiter = EntityType.extractRules(entityType, duration, getMailetConfig)
-      .map(rateLimiterFactory.withSpecification(_, precision)),
-      keyPrefix = keyPrefix,
-      entityType = entityType)
+    EntityType.extractRules(entityType, duration, getMailetConfig)
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala:151]: :+1:

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala [hunk 8]
```diff
@@ -56,21 +56,32 @@ object ConfigurationOps {
 }
 
 object EntityType {
+
+  implicit class EitherOps[E <: Throwable, A](either: Either[E, A]) {
```
Comment (jeantil) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:60]: this one I agree with :+1:

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala (hunk 4 above)
Comment (mbaechler) [reply server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:64]: Good question (I don't have the answer). What do you think @chibenwa @Arsnael ?

I would also add that if you put this concern aside, the code is usually more cohesive with fp style.

Comment (chibenwa) [reply server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:64]: > Good question (I don't have the answer). What do you think @chibenwa @Arsnael ?

I expect it to be mostly table but maybe more entity types (HeaderCount? Maybe more inventive people that me can find new ideas?)

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala [hunk 9]
```diff
@@ -48,62 +49,48 @@ object ConfigurationOps {
...
+        case s  => s
       }
       .map(org.apache.james.util.Size.parse)
   }
 }
 
+object EntityType {
+  def extractRules(entityType: EntityType, duration: Duration, mailetConfig: MailetConfig): Option[Rules] = (entityType match {
+    case Count          => mailetConfig.getOptionalLong("count")
+    case RecipientsType => mailetConfig.getOptionalLong("recipients")
```
Comment (chibenwa) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:66]:
```suggestion
    case Recipients => mailetConfig.getOptionalLong("recipients")
```

While we are at it?

**Review by chibenwa — APPROVED** [review]

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala (hunk 4 above)
Comment (Arsnael) [reply server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/EntityType.scala:64]: > Good question (I don't have the answer). What do you think @chibenwa @Arsnael ?

I don't really have a proper answer either but I'm ok with the change

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
@jeantil are you ok with those changes? Did it answer your remarks? :)

**Pull request merged**